types/validator_set.go

```diff
@@ -680,28 +681,69 @@ func (vals *ValidatorSet) VerifyCommit(chainID string, blockID BlockID,
 
 	talliedVotingPower := int64(0)
 	votingPowerNeeded := vals.TotalVotingPower() * 2 / 3
-	for idx, commitSig := range commit.Signatures {
-		if commitSig.Absent() {
-			continue // OK, some signatures can be absent.
-		}
+	bv, batchVerify := batch.CreateBatchVerifier(vals.GetProposer().PubKey)
+	if batchVerify && len(commit.Signatures) > 1 {
+		for idx, commitSig := range commit.Signatures {
+			if commitSig.Absent() {
+				continue // OK, some signatures can be absent.
+			}
 
-		// The vals and commit have a 1-to-1 correspondance.
-		// This means we don't need the validator address or to do any lookup.
-		val := vals.Validators[idx]
+			// The vals and commit have a 1-to-1 correspondance.
+			// This means we don't need the validator address or to do any lookup.
+			val := vals.Validators[idx]
+
+			// Validate signature.
+			voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
+			// add the key, sig and message to the verifier
+			if err := bv.Add(val.PubKey, voteSignBytes, commitSig.Signature); err != nil {
+				return err
+			}
 
-		// Validate signature.
-		voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
-		if !val.PubKey.VerifySignature(voteSignBytes, commitSig.Signature) {
-			return fmt.Errorf("wrong signature (#%d): %X", idx, commitSig.Signature)
+			// Good!
+			if commitSig.ForBlock() {
+				talliedVotingPower += val.VotingPower
+			}
 		}
-		// Good!
-		if commitSig.ForBlock() {
-			talliedVotingPower += val.VotingPower
+		if !bv.Verify() {
+			talliedVotingPower = 0
+			for idx, commitSig := range commit.Signatures {
+				if commitSig.Absent() {
+					continue // OK, some signatures can be absent.
+				}
+
+				val := vals.Validators[idx]
+				voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
+				if !val.PubKey.VerifySignature(voteSignBytes, commitSig.Signature) {
+					return fmt.Errorf("wrong signature (#%d): %X", idx, commitSig.Signature)
+				}
+
+				// Good!
+				if commitSig.ForBlock() {
+					talliedVotingPower += val.VotingPower
+				}
+			}
+		}
+	} else {
+		for idx, commitSig := range commit.Signatures {
+			if commitSig.Absent() {
+				continue // OK, some signatures can be absent.
+			}
+
+			// The vals and commit have a 1-to-1 correspondance.
+			// This means we don't need the validator address or to do any lookup.
+			val := vals.Validators[idx]
+
+			// Validate signature.
+			voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
+			if !val.PubKey.VerifySignature(voteSignBytes, commitSig.Signature) {
+				return fmt.Errorf("wrong signature (#%d): %X", idx, commitSig.Signature)
+			}
+
+			// Good!
+			if commitSig.ForBlock() {
+				talliedVotingPower += val.VotingPower
+			}
```

Comment:
So this block and the block under the case of when `!bv.Verify()` evaluates true are identical. We should keep this code DRY. Perhaps move this logic to an inline function in this function.

e.g.

```go
verifySigsDefault := func(...) {
  // ...
}

// ...
if !bv.Verify() {
  verifySigsDefault(...)
}
```

Also, if `!bv.Verify()` evaluates to true, we recompute the `voteSignBytes` all over again. Maybe this isn't much overhead, but perhaps we can compute and store these once. Thoughts @melekes?

Reply:
I can look at inlining, I tested it and ran into some issues but cant recall why I reverted. 

We can try caching it. If verification fails we will need to single verify whats in the cache and potentially more. Its possible to do this, I can do some benchmarking to test this.

Reply:
It should 100% be possible. The code is identical unless I'm missing something obscure.

Reply:
agree re code duplication